src/progress_bar.rs

```diff
@@ -643,131 +622,16 @@ impl MultiProgress {
-                let split = draw_state.lines.split_at(draw_state.orphan_lines);
-                orphan_lines.extend_from_slice(split.0);
-                split.1.to_vec()
-            } else {
-                draw_state.lines
-            };
-
-            let draw_state = ProgressDrawState {
-                lines,
-                orphan_lines: 0,
-                ..draw_state
-            };
-
-            if let Some(ref mut obj) = &mut state.objects[idx] {
-                obj.draw_state = Some(draw_state);
-            }
-
-            // the rest from here is only drawing, we can skip it.
-            if state.draw_target.is_hidden() {
-                continue;
-            }
-
-            debug_assert!(recv_peek.is_none());
-            if grouped >= MAX_GROUP_SIZE {
-                // Can't group any more draw calls, proceed to just draw
-                grouped = 0;
-            } else if let Ok(state) = self.rx.try_recv() {
-                // Only group draw calls if there is another draw already queued
-                recv_peek = Some(state);
-                grouped += 1;
-                continue;
-            } else {
-                // No more draws queued, proceed to just draw
-                grouped = 0;
-            }
-
-            let mut lines = vec![];
-
-            // Make orphaned lines appear at the top, so they can be properly
-            // forgotten.
-            let orphan_lines_count = orphan_lines.len();
-            lines.append(&mut orphan_lines);
-
-            for index in state.ordering.iter() {
-                if let Some(obj) = &state.objects[*index] {
-                    if let Some(ref draw_state) = obj.draw_state {
-                        lines.extend_from_slice(&draw_state.lines[..]);
-                    }
-                }
-            }
-
-            let finished = state.is_done();
-            state.draw_target.apply_draw_state(ProgressDrawState {
-                lines,
-                orphan_lines: orphan_lines_count,
-                force_draw: force_draw || orphan_lines_count > 0,
-                move_cursor,
-                finished,
-            })?;
-
-            force_draw = false;
-        }
-
-        if clear {
-            let mut state = self.state.write().unwrap();
-            state.draw_target.apply_draw_state(ProgressDrawState {
-                lines: vec![],
-                orphan_lines: 0,
-                finished: true,
-                force_draw: true,
-                move_cursor,
-            })?;
-        }
-
-        self.joining.store(false, Ordering::Release);
+    pub fn clear(&self) -> io::Result<()> {
+        let mut state = self.state.write().unwrap();
+        let move_cursor = state.move_cursor;
+        state.draw_target.apply_draw_state(ProgressDrawState {
+            lines: vec![],
+            orphan_lines: 0,
+            finished: true,
+            force_draw: true,
+            move_cursor,
+        })?;
```

Comment:
It would be nice to extract this method from `join_impl()` in a separate commit before this one, so it becomes clearer how this changed.